Vwitter/SpotifyAPI/SpotifyAPIManager.m

```diff
@@ -0,0 +1,212 @@
+}
+
+- (BOOL)hasValidSpotifyAuthorization {
+    NSTimeInterval timestampSeconds = NSDate.date.timeIntervalSince1970;
+    return (self.accessToken != nil
+            && self.accessTokenExpirationTimestampSeconds != nil
+            && self.accessTokenExpirationTimestampSeconds.doubleValue > timestampSeconds);
+}
+
+- (void)setupAppRemote {
+    /*
+     Scopes let you specify exactly what types of data your application wants to
+     access, and the set of scopes you pass in your call determines what access
+     permissions the user is asked to grant.
+     For more information, see https://developer.spotify.com/web-api/using-scopes/.
+     */
+    SPTScope scope = SPTUserLibraryReadScope | SPTPlaylistReadPrivateScope;
+
+    /*
+     Start the authorization process. This requires user input.
+     */
+    if (@available(iOS 11, *)) {
+        // Use this on iOS 11 and above to take advantage of SFAuthenticationSession
+        [self.sessionManager initiateSessionWithScope:scope options:SPTDefaultAuthorizationOption];
+        NSLog(@"authorized spotify session");
+    } else {
+        // Use this on iOS versions < 11 to use SFSafariViewController
+        NSLog(@"iOS version too old");
+    }
+    
+    SPTConfiguration *configuration =
+        [[SPTConfiguration alloc] initWithClientID:SpotifyClientID redirectURL:[NSURL URLWithString:SpotifyRedirectURLString]];
+
+    self.appRemote = [[SPTAppRemote alloc] initWithConfiguration:configuration logLevel:SPTAppRemoteLogLevelDebug];
+}
+
+- (void)authorizeSpotify {
+    [self setupAppRemote];
+    [self.appRemote authorizeAndPlayURI:@"spotify:artist:2YZyLoL8N0Wb9xBt1NhZWg"];
+}
+
+- (void)playTrack:(NSString *)trackUri {
+    if (!self.hasValidSpotifyAuthorization) {
+        [self authorizeSpotify];
+    }
+    
+    if (!self.appRemote.playerAPI) {
+        [self.appRemote connect];
+    }
+    
+    [self.appRemote.playerAPI play:trackUri callback:^(id  _Nullable result, NSError * _Nullable error) {
+        if (!error) {
+            NSLog(@"track playing");
+        }
+        else {
+            NSLog(@"cell track failed to play %@", error.localizedDescription);
+        }
+    }];
+}
+
+- (void)pause {
+    [self.appRemote.playerAPI pause:^(id  _Nullable result, NSError * _Nullable error) {
+        if (!error) {
+            NSLog(@"track paused");
+
+        }
+        else {
+            NSLog(@"track not paused, %@", error.localizedDescription);
+        }
+    }];
+    
+}
+
+- (void)getTracks:(NSString *)searchString withCompletion:(SpotifyTrackCompletion)completion {
+    if (!self.hasValidSpotifyAuthorization) {
+        [self authorizeSpotify];
+    }
+    
+    NSString *bearerToken = [[NSString alloc] initWithFormat:@"Bearer %@", self.appRemote.connectionParameters.accessToken];
+    
+    NSMutableURLRequest *request = [[NSMutableURLRequest alloc] init];
+    
+    NSString *convertedSearchString = [searchString stringByReplacingOccurrencesOfString:@" " withString:@"%20"];
```

Comment:
loved seeing this, but it is still somewhat incorrect. The problem is, URL's cannot have things like spaces in them but there are other bad characters as well, and so the standard practice is to use what is called "URL encoding" to replace invalid things like spaces into something like %20 which every system should understand is a space.

I would recommend doing a search on stackoverflow to find a simpler way to do URL encoding of an NSString, it should be pretty easy to handle basically anything beyond just spaces